File: 프로그래머스/lv2/87946. 피로도/피로도.java
class Solution {
    public int solution(int k, int[][] dungeons) {
        boolean [] visited = new boolean[dungeons.length];
        int answer = dfs (k, dungeons, visited, 0, 0);

        return answer;
    }
    
    private int dfs(int k, int[][] dungeons, boolean [] visited, int maxCnt, int cnt){
        int m = maxCnt;
        for (int i=0; i< dungeons.length; i++){
            if (!visited[i] && k >= dungeons[i][0]){
                visited[i] = true;
                m = dfs(k - dungeons[i][1], dungeons, visited, m, cnt +1);
                visited[i] = false;
            }
        }
        return Math.max(cnt, m);
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    def dfs(k, dungeons, visited, cnt):
        max_cnt = cnt
        for i in range(len(dungeons)):
            if not visited[i] and k >= dungeons[i][0]:
                visited[i] = True
                max_cnt = max(max_cnt, dfs(k - dungeons[i][1], dungeons, visited, cnt + 1))
                visited[i] = False
        return max_cnt

    visited = [False] * len(dungeons)
    answer = dfs(k, dungeons, visited, 0)
    return answer